src/as3compile.1
.TH as3compile "1" "April 2012" "as3compile" "swftools"
.SH NAME
as3compile \- Compile .as ActionScript 3.0 files to swf.

.SH Synopsis
.B as3compile file.as [\-o file.swf] 

.SH DESCRIPTION
Compiles a file written in ActionScript 3.0 to a SWF file.

.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR 
    Print short help message and exit
.TP
\fB\-V\fR, \fB\-\-version\fR 
    Print version info and exit
.TP
\fB\-v\fR, \fB\-\-verbose\fR 
    Increase verbosity
.TP
\fB\-q\fR, \fB\-\-quiet\fR 
    Decrease verbosity
.TP
\fB\-C\fR, \fB\-\-cgi\fR 
    Output to stdout (for use in CGI environments)
.TP
\fB\-R\fR, \fB\-\-resolve\fR 
    This flag will cause the compiler to try filenames like "FooBar.as" for classes named "FooBar".
.TP
\fB\-D\fR, \fB\-\-define\fR \fInamespace::variable\fR
    Set a compile time variable (for doing conditional compilation)
.TP
\fB\-X\fR, \fB\-\-width\fR 
    Set target SWF width
.TP
\fB\-Y\fR, \fB\-\-height\fR 
    Set target SWF width
.TP
\fB\-r\fR, \fB\-\-rate\fR 
    Set target SWF framerate
.TP
\fB\-M\fR, \fB\-\-mainclass\fR 
    Set the name of the main class (extending flash.display.MovieClip)
.TP
\fB\-l\fR, \fB\-\-library\fR \fIfile\fR
    Include library file \fIfile\fR. <file> can be an .abc or .swf file.
.TP
\fB\-I\fR, \fB\-\-include\fR \fIdir\fR
    Add additional include dir \fIdir\fR.
.TP
\fB\-N\fR, \fB\-\-local-with-network\fR 
    Make output file "local with networking"
.TP
\fB\-L\fR, \fB\-\-local-with-filesystem\fR 
    Make output file "local with filesystem"
.TP
\fB\-T\fR, \fB\-\-flashversion\fR \fInum\fR
    \fInum\fR must be >= 9.
.TP
\fB\-o\fR, \fB\-\-output\fR \fIfilename\fR
    Set output file to \fIfilename\fR.
.SH EXAMPLE

 The following is a basic as3 file that can be compiled e.g.
 with \fBas3compile \-X 400 \-Y 400 smiley.as\fR

 // smiley.as
 package
 {
     import flash.display.MovieClip
 
     public class Main extends MovieClip
     {
         function Main()
         {
             this.graphics.beginFill(0xcccc00)
             this.graphics.drawCircle(200,200,200)
             this.graphics.endFill()
             this.graphics.beginFill(0x000000)
             this.graphics.drawCircle(140,150,50)
             this.graphics.drawCircle(260,150,50)
             this.graphics.drawRoundRect(140,270,120,10,20);
             this.graphics.endFill()
         }
     }
 }

.SH AUTHOR

Matthias Kramm <kramm@quiss.org>


ChangeLog
0.0.1: initial release
0.0.2: 
	* pdf2swf: Added support for pdf bitmaps, clipping, state saving/restoring, colors,
	           solid fill and line width. The output SWF files are now much smaller.
	* swfcombine: Added stacking option (-t)
	* swfcombine: Small bugfixes with the command line options
	* rfxswf-lib: some bugfixes concerning sprites, cgi, shapes, jpegs.
	* swfstrings: New: Allows extraction of text from swf files
	* swfdump: New: Dumps swf file information. (And does some error checking on the SWF)
	* jpeg2swf: New: Converts one or more jpegs to a swf movie.

0.1.0: 
	* configure: Now works without automake, automatically disables 
			jpeg2swf and pdf2swf if jpeglib is not found (disable
			both) or t1lib is not found (disable pdf2swf)
	* swfcombine: bugfix: DEFINETEXT tags didn't get relocated right
	* swfcombine: bugfix: no longer builds "unstreamable" SWFs
	* rfxswf-lib: actionscript
	* rfxswf-lib: lossless image formats
	* pdf2swf: conversion of lossless image formats
	* pdf2swf: picture cache to prevent images getting stored more than once
	* pdf2swf: conversion of (hyper)links
	* swfdump: -a, -e and -XY options
	* lib/example/zlibtest: new file, demonstrates using lossless images with rfxswf-lib
	* lib/example/dumpfont: utility for dumping fonts out of swf (and generate c sourcecode for creating them)
	* lib/example/demofont: example font, generated with lib/example/dumpfont.

0.1.1:  
	* swfcombine: Wasn't able to handle converted pdfs with links. Fixed.
	* swfcombine: Introduce --dummy, -X,-Y,-r options
	* swfdump: new: -r option
	* pdf2swf: small bugfixes
	* HOWTO_pdf2swf: added section about correcting size and framerate

0.1.2:
	* swfdump: can dump all actionscript flavours (button, button2, standalone) now
	* rfxswf-lib: improved font handling
	* lib/example/makefonts.cc: new file, converts the standard fonts to swf
	* lib/example/glyphshape.c: new file, demonstrates using font characters as shapes
	* pdfswf: many bugfixes. The "pdfscreen" manual now get's converted correctly
	* rfxswf-lib: bugfix: Couldn't handle jpeg grayscale images

0.2.0:	
	* pdf2swf: TrueType support added. (using ttf2pt1)
	* rfxswf-lib: performs conversion of jpeg colorspaces now. (Makes pdf2swf cope
	               with more types of jpegs)
	* swfextract: new tool, allows to extract single objects, frames etc. from movieclips
	* swfcombine: allows ids instead of names, now, too.
	* swfdump: added text displaying functionality. (like swfstrings)
	* swfdump: now dumps also the undocumented placeobject2 actioncode.
	* lib/example/avi2swf.cc: For now, converts small (!) avi movie files to swf.  (using 
			          libavifile)

0.2.1:	
	* configure: Does checking for byte order and command substitution capability now.
	* rfxswf: Applied Anjo Krank's patch for MacOS X.
	* pdf2swf: Fixed font transformation matrix
	* swfcombine: new options: --cat, --stack1, --merge, --frame
	* swfs/: new directory, holds some standard SWFs primarily used for pdf2swf
	         installs into ${DATADIR}/swftools/swfs
	* pdf2swf: New options --viewer, --preloader, -b, -l
	* pdf2swf: Fonts are now in ${DATADIR}/swftools/fonts

0.2.2:	
	* swfs/: Added/Wrote sources for some of the swf files 
	* swfdump: Major bugfix in placeobject2, hex output function
	* swfextract: Can now extract png, jpeg and mp3.
	* swfcombine: big endian fixes
	* swfextract: big endian fixes
	* pdf2swf: Changed from 100dpi to 72dpi conversion
	* pdf2swf: Some optimizations, fix the zero string problem.
	* lib/: Added Bladeenc 0.94.1 for mp3 compression
	* lib/example/avi2swf.cc: Can now handle big files. Slightly better compression ratio. Sound.
	* lib/example/sound.c: New file, demonstrates use of the sound module in librfxswf.

0.2.3: 
	* pdf2swf: New option --fontdir, can now handle external fonts.
	* lib/modules/swftext.c: Several font-extraction bugfixes
	* src/swfdump: Bugfixes, ids and depths are now in decimal
	* src/swfcombine: Bugfixes.
	* configure.in: Cleaned up.

0.3.0:
	* pdf2swf: The --ignore option now actually makes the pdf files smaller.
	* rfxswf: Can now write DefineFont2 tags
	* swfextract/swfcombine: Bugfixes (DefineEditText)
	* swfdump: New option --used, prints out referred IDs of each tag.
	* rfxswf: Can now read and write Flash 6 (MX) compressed SWFs. (SWCs)
	* pdf2swf: Added compression flag (-z)
	* swfcombine: Added compression flag (-z)
	* configure.in: zlib is now required.
	* src/png2swf: New tool.
	* swfdump: Prints some image infos now.
	* pdf2swf: Fixed Image handling
	* avi2swf/: Is part of the main distribution now
	* src/wav2swf: New tool.

0.3.1:
	* configure: Didn't create all Makefiles when t1lib or jpeglib were absent. Fixed.
	* swfs/Makefile: FreeBSD fixes.

0.4.0:
	* pdf2swf/xpdf: Upgraded PDF Parser to Version 1.01.
	* pdf2swf: Can now handle Type 3 Fonts. (As Bitmaps)
	* src/swfcombine: Completely rewritten, using the rfxswf routines
	* configure: Check for libtool
	* lib/rfxswf: Fixed some parsing bugs (swffont, swfdump)
	* pdf2swf: Added antialiasing capabilities for Type 3 Fonts
	* lib/examples/makefonts: Fonts: Bounding boxes are now handled correctly
	* png2swf: Support for more image types
	* pdf2swf: New option -t (--stop)

0.4.1:
	* src/swfcombine: Bugfix with -x option
	* lib/Makefile: MacOSX fixes
	* pdf2swf/SWFOutput.cc: fixed undefined sqrt()
	* src/png2swf: Support for transparency
	* pdf2swf: Support for CID Fonts
	* src/swfcombine: Bugfixes
	* lib/rfxswf: Added Sprite Unfolding routines

0.4.2:
	* src/Makefile: OpenBSD fixes
	* lib/lame: Removed *.o files from archive
	* avi2swf: Implemented --flip option
	* pdf2swf: Implemented --flashversion option
	* pdf2swf: Implemented Viewer gadgets- it's now possible to call Flash functions via hyperlink

0.4.3:
	* src/jpeg2swf: Grayscale fix
	* lib/rfxswf: Improved MX support
	* pdf2swf: Command line fixes
	* swfdump: Added -f (--frames) option
	* png2swf: Added support for mode 6 images
	* swfs: Support for print protection in simple_viewer

src/as3compile.c
/* as3compile.c
   Compiles ActionScript 3.0 (.as) files into .swf files.

   Part of the swftools package.

   Copyright (c) 2008/2009 Matthias Kramm <kramm@quiss.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include "../lib/rfxswf.h"
#include "../lib/args.h"
#include "../lib/q.h"
#include "../lib/os.h"

static char * filename = 0;
static char * outputname = 0;
static int override_outputname = 0;
static int do_cgi = 0;
static double framerate = 25.0;
static double width = 400;
static double height = 300;
static int flashversion = 9;
static int verbose = 1;
static char local_with_networking = 0;
static char local_with_filesystem = 0;
static char*mainclass = 0;

static struct options_t options[] = {
{"h", "help"},
{"V", "version"},
{"v", "verbose"},
{"q", "quiet"},
{"C", "cgi"},
{"R", "resolve"},
{"D", "define"},
{"X", "width"},
{"Y", "height"},
{"r", "rate"},
{"M", "mainclass"},
{"l", "library"},
{"I", "include"},
{"N", "local-with-network"},
{"L", "local-with-filesystem"},
{"T", "flashversion"},
{"o", "output"},
{0,0}
};

int args_callback_option(char*name,char*val)
{
    if(!strcmp(name, "V")) {
        printf("swfc - part of %s %s\n", PACKAGE, VERSION);
        exit(0);
    }
    else if(!strcmp(name, "o")) {
	outputname = val;
	override_outputname = 1;
	return 1;
    }
    else if(!strcmp(name, "r")) {
	framerate = atof(val);
        return 1;
    }
    else if(!strcmp(name, "M")) {
        mainclass = val;
        return 1;
    }
    else if(!strcmp(name, "v")) {
	verbose++;
        return 0;
    }
    else if(!strcmp(name, "q")) {
	verbose--;
        return 0;
    }
    else if(!strcmp(name, "X")) {
	width = atof(val);
        return 1;
    }
    else if(!strcmp(name, "Y")) {
	height = atof(val);
        return 1;
    }
    else if(!strcmp(name, "T")) {
        flashversion = atoi(val);
        return 1;
    }
    else if(!strcmp(name, "C")) {
	do_cgi = 1;
	return 0;
    }
    else if(!strcmp(name, "l")) {
        as3_import_file(val);
	return 1;
    }
    else if(!strcmp(name, "I")) {
        as3_add_include_dir(val);
	return 1;
    }
    else if(!strcmp(name, "R")) {
        as3_set_option("recurse","1");
	return 0;
    }
    else if(!strcmp(name, "D")) {
        if(!strstr(val, "::")) {
            fprintf(stderr, "Error: compile definition must contain \"::\"\n");
            exit(1);
        }
        as3_set_define(val);
	return 1;
    }
    else if (!strcmp(name, "N"))
    {
	local_with_networking = 1;
	return 0;
    }
    else if (!strcmp(name, "L"))
    {
	local_with_filesystem = 1;
	return 0;
    }
    else {
        printf("Unknown option: -%s\n", name);
	exit(1);
    }
    return 0;
}
int args_callback_longoption(char*name,char*val)
{
    return args_long2shortoption(options, name, val);
}
void args_callback_usage(char *name)
{
    printf("\n");
    printf("Usage: %s file.as [-o file.swf] \n", name);
    printf("\n");
    printf("-h , --help                    Print short help message and exit\n");
    printf("-V , --version                 Print version info and exit\n");
    printf("-v , --verbose                 Increase verbosity\n");
    printf("-q , --quiet                   Decrease verbosity\n");
    printf("-C , --cgi                     Output to stdout (for use in CGI environments)\n");
    printf("-R , --resolve                 Try to resolve undefined classes automatically.\n");
    printf("-D , --define <namespace::variable>    Set a compile time variable (for doing conditional compilation)\n");
    printf("-X , --width                   Set target SWF width\n");
    printf("-Y , --height                  Set target SWF width\n");
    printf("-r , --rate                    Set target SWF framerate\n");
    printf("-M , --mainclass               Set the name of the main class (extending flash.display.MovieClip)\n");
    printf("-l , --library <file>          Include library file <file>. <file> can be an .abc or .swf file.\n");
    printf("-I , --include <dir>           Add additional include dir <dir>.\n");
    printf("-N , --local-with-network      Make output file \"local with networking\"\n");
    printf("-L , --local-with-filesystem     Make output file \"local with filesystem\"\n");
    printf("-T , --flashversion <num>      Set target SWF flash version to <num>.\n");
    printf("-o , --output <filename>       Set output file to <filename>.\n");
    printf("\n");
}
int args_callback_command(char*name,char*val)
{
    if(filename) {
        fprintf(stderr, "Only one file allowed. You supplied at least two. (%s and %s)\n",
                 filename, name);
    }
    filename = name;
    return 0;
}

void writeSWF(SWF*swf)
{
    int fi = -1;
    if(do_cgi || !strcmp(filename, "-"))
	fi = fileno(stdout);
    else
	fi = open(outputname, O_WRONLY|O_CREAT|O_TRUNC|O_BINARY, 0644);
    if(fi<0) {
	fprintf(stderr, "couldn't create output file %s\n", filename);
        exit(1);
    }
    if(do_cgi) {
        if(swf_WriteCGI(swf)<0) {
            fprintf(stderr, "WriteCGI failed.\n");
            exit(1);
        }
    } else {
	if(swf_WriteSWF(fi, swf)<0) {
            fprintf(stderr, "WriteSWF() failed.\n");
            exit(1);
        }
    }
}

int main (int argc,char ** argv)
{
    char buf[512];
    char*currentdir = getcwd(buf, 512);
    if(!currentdir) {
        as3_warning("Could not determine the current directory");
    } else {
        as3_add_include_dir(currentdir);
    }

    int t;
    processargs(argc, argv);
    as3_setverbosity(verbose);

    if(!filename) {
	args_callback_usage(argv[0]);
	exit(1);
    }
    if(!outputname) {
	outputname = stripFilename(filename, ".swf");
        //as3_warning("output name not given, writing to %s", outputname);
    }

    if(!strcmp(filename, ".")) {
        as3_parse_directory(".");
    } else {
        as3_parse_file(filename);
    }

    void*code = as3_getcode();

    SWF swf;
    memset(&swf, 0, sizeof(swf));
    swf.fileVersion = flashversion;
    swf.frameRate = framerate*0x100;
    swf.movieSize.xmin = 0;
    swf.movieSize.ymin = 0;
    swf.movieSize.xmax = width*20;
    swf.movieSize.ymax = height*20;
    TAG*tag = swf.firstTag = swf_InsertTag(0, ST_DOABC);
    swf_WriteABC(tag, code);

    if(!mainclass)
        mainclass = as3_getglobalclass();
    if(mainclass) {
        tag = swf_InsertTag(tag, ST_SYMBOLCLASS);
        swf_SetU16(tag, 1);
        swf_SetU16(tag, 0);
        swf_SetString(tag, mainclass);
    } else {
        as3_warning("no global public MovieClip subclass");
    }
    
    as3_destroy();

    tag = swf_InsertTag(tag, ST_SHOWFRAME);
    tag = swf_InsertTag(tag, ST_END);

    swf_FreeABC(code);
    
    if(local_with_filesystem)
        swf.fileAttributes &= ~FILEATTRIBUTE_USENETWORK;
    if(local_with_networking)
        swf.fileAttributes |= FILEATTRIBUTE_USENETWORK;

    writeSWF(&swf);
    swf_FreeTags(&swf);

    return 0;
}

